File: my_py_pkg/my_py_pkg/number_counter.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node

from example_interfaces.msg import Int64
from example_interfaces.srv import SetBool

class NumberCounterNode(Node): # MODIFY NAME
    def __init__(self):
        super().__init__("number_counter") # MODIFY NAME
        self.count = 1
        self.total = 0
        self.message = "No message received yet"
        self.subscriber_ = self.create_subscription(Int64, "number", self.callback_number_counter, 10)
        self.publisher_ = self.create_publisher(Int64, "number_count", 10)
        self.server_ = self.create_service(SetBool, "reset_counter", self.callback_set_bool)
        self.get_logger().info("Number counter has been started")
        self.publish_num()
    
    def callback_number_counter(self, msg):
        self.message = f"Received \"{msg.data}\" {self.count} times"
        self.get_logger().info(self.message)
        self.count += 1
        self.total += msg.data
        self.publish_num()
    
    def publish_num(self):
        msg = Int64()
        msg.data = self.total
        self.publisher_.publish(msg)
    
    def callback_set_bool(self, request, response):
        if request.data:
            response.success = True
            response.message = "Counter reset"
            self.total = 0
        else:
            response.success = False
            response.message = "Counter not reset"
        return response
        
 
def main(args=None):
    rclpy.init(args=args)
    node = NumberCounterNode() # MODIFY NAME
    rclpy.spin(node)
    rclpy.shutdown()
 
 
if __name__ == "__main__":
    main()

File: my_cpp_pkg/src/number_counter.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/msg/int64.hpp"
#include "example_interfaces/srv/set_bool.hpp"

using std::placeholders::_1;
using std::placeholders::_2;

class NumberCounterNode : public rclcpp::Node {
public:
    NumberCounterNode() : Node("number_counter") {
        total_ = 0;
        subscriber_ = this->create_subscription<example_interfaces::msg::Int64>("number", 10,
                                    std::bind(&NumberCounterNode::callbackNumber, this, std::placeholders::_1));
        publisher_ = this->create_publisher<example_interfaces::msg::Int64>("number_count", 10);
        server_ = this->create_service<example_interfaces::srv::SetBool>("reset_counter",
                    std::bind(&NumberCounterNode::callbackResetNumber, this, _1, _2));
        RCLCPP_INFO(this->get_logger(), "Number counter has been started");
    }

private:
    void callbackNumber(const example_interfaces::msg::Int64::SharedPtr msg) {
        total_+= msg->data;
        RCLCPP_INFO(this->get_logger(), "Received %ld, the total is now %d.", msg->data, total_);
        auto num = example_interfaces::msg::Int64();
        num.data = total_;
        publisher_->publish(num);
    }

    void callbackResetNumber(const example_interfaces::srv::SetBool::Request::SharedPtr request,
                             const example_interfaces::srv::SetBool::Response::SharedPtr response) {
        if (request->data) {
            total_ = 0;
            response->success = true;
            response->message = "Counter reset";
        }
        else {
            response->success = false;
            response->message = "Counter not reset";
        }
    }

    int total_;
    rclcpp::Subscription<example_interfaces::msg::Int64>::SharedPtr subscriber_;
    rclcpp::Publisher<example_interfaces::msg::Int64>::SharedPtr publisher_;
    rclcpp::Service<example_interfaces::srv::SetBool>::SharedPtr server_;
};


int main(int argc, char **argv) {
    rclcpp::init(argc, argv);
    auto node = std::make_shared<NumberCounterNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}